src/Microsoft.AspNet.Server.Kestrel/Http/SocketOutput.cs

```diff
@@ -138,6 +144,58 @@ public void End(ProduceEndType endType)
             }
         }
 
+        public MemoryPoolIterator2 ProducingStart()
+        {
+            lock (_returnLock)
+            {
+                Debug.Assert(!_isProducing);
+                _isProducing = true;
+
+                if (_tail == null)
+                {
+                    throw new IOException("The socket has been closed.");
+                }
+
+                return new MemoryPoolIterator2(_tail, _tail.End);
+            }
+        }
+
+        public void ProducingComplete(MemoryPoolIterator2 end, int count)
```

Comment:
Then its just appended to the end of the current chain, if it hasn't been sent yet.